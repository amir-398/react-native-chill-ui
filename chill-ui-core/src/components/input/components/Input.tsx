import { InputProps } from '@types';
import { Box } from '@components/box';
import { Icon } from '@components/icon';
import { String } from '@components/string';
import { AnimatedBox } from '@components/animatedBox';
import { Animated, Pressable, TextInput } from 'react-native';
import { forwardRef, useEffect, useMemo, useRef, useState } from 'react';
import { cn, getStringLength, isString, classNamePropsHandler, classNameHandler, styleHandler } from '@utils';

import { inputDefaultProps } from '../utils/defaultProps';
import { inputContainerSv, inputSv, styles } from '../styles/Input.ss.styles';
import { inputContainerTv, inputFieldTv, twStyles } from '../styles/Input.tw.styles';

/**
 * Input component with Hybrid styling (Tailwind + StyleSheet).
 * Provides a comprehensive text input with validation, icons, error handling, and customizable styling.
 * Automatically detects NativeWind availability and falls back to StyleSheet if needed.
 *
 * @example
 * ```tsx
 * // Basic input with label
 * <Input
 *   label="Email"
 *   placeholder="Enter your email"
 *   value={email}
 *   onChangeText={setEmail}
 * />
 *
 * // Input with validation and error handling
 * <Input
 *   label="Password"
 *   placeholder="Enter password"
 *   hasSecureTextEntry
 *   hasError={hasError}
 *   errorMessage="Password is required"
 *   allow="lettersWithoutSpecialCharacters"
 * />
 * ```
 *
 * @param allow - Input validation type ('all' | 'numbers' | 'letters' | 'lettersWithoutSpecialCharacters')
 * @param className - Custom CSS classes for the input container
 * @param clickableAs - Animation type when input is pressed ('scale' | undefined)
 * @param customRegex - Custom regex pattern for input validation
 * @param editable - Whether the input is editable (default: true)
 * @param errorClassName - Custom CSS classes for error state
 * @param errorStyle - Custom CSS classes for the error message
 * @param inputStyle - Custom CSS classes for the input field
 * @param errorIconName - Icon name to display with error message
 * @param errorMessage - Error message to display below input
 * @param errorStringProps - Props for the error message String component
 * @param hasClearIcon - Whether to show clear icon when input has value (default: true)
 * @param hasError - Whether input is in error state
 * @param hasSecureTextEntry - Whether input should hide text (password field)
 * @param inputClassName - Custom CSS classes for the input field
 * @param isDisabled - Whether input is disabled
 * @param isStretchable - Whether input should stretch to full width
 * @param label - Label text to display above input
 * @param labelStringProps - Props for the label String component
 * @param leftIconAction - Configuration for left icon (iconName, iconSize, iconColor, iconPress, customIcon)
 * @param lengthStringProps - Props for the character count String component
 * @param maxLength - Maximum number of characters allowed
 * @param multiline - Whether input supports multiple lines
 * @param onChangeText - Callback when input text changes
 * @param onPress - Callback when input is pressed
 * @param rightIconAction - Configuration for right icon (iconName, iconSize, iconColor, iconPress, customIcon)
 * @param showLength - Whether to show character count
 * @param size - Input size variant ('xs' | 'sm' | 'md' | 'lg' | 'xl')
 * @param value - Current input value
 * @param wrapperRef - Ref for the input container wrapper
 * @returns Styled input component with validation and icon support
 */
const Input = forwardRef<TextInput, InputProps>((props, ref) => {
  classNamePropsHandler(props, 'Input');
  const {
    allow = inputDefaultProps.allow,
    className,
    clickableAs,
    customRegex,
    editable = inputDefaultProps.editable,
    errorClassName,
    errorIconName,
    errorMessage,
    errorStringProps,
    errorStyle,
    hasClearIcon = inputDefaultProps.hasClearIcon,
    hasError,
    hasSecureTextEntry,
    inputClassName,
    inputStyle,
    isDisabled,
    isStretchable,
    label,
    labelStringProps,
    leftIconAction,
    lengthStringProps,
    maxLength,
    multiline,
    onChangeText,
    onPress,
    rightIconAction,
    showLength,
    size = inputDefaultProps.size,
    style,
    value,
    wrapperRef,
    ...rest
  } = props;

  const [isSecureEntry, setIsSecureEntry] = useState<boolean>(hasSecureTextEntry || false);
  const [inputValue, setInputValue] = useState<string>(value || '');
  const scaleAnim = useRef(new Animated.Value(1)).current;

  const xmarkIconSize = useMemo(() => {
    if (size === 'lg' || size === 'xl') return 'sm';
    return 'xs';
  }, [size]);

  /**
   * Validates input text based on allow prop or custom regex
   */
  const validateInput = (text: string): boolean => {
    if (customRegex) {
      return customRegex.test(text);
    }

    switch (allow) {
      case 'numbers':
        return /^\d*$/.test(text);
      case 'letters':
        return /^[a-zA-ZÀ-ÿ\s]*$/.test(text);
      case 'lettersWithoutSpecialCharacters':
        return /^[a-zA-ZÀ-ÿ0-9\s]*$/.test(text);
      case 'all':
      default:
        return true;
    }
  };

  /**
   * Clears the input value and calls onChangeText callback
   */
  const handleClearInput = () => {
    setInputValue('');
    onChangeText?.('');
  };

  /**
   * Handles text changes with validation and state updates
   */
  const handleOnChange = (text: string) => {
    if (validateInput(text)) {
      !isString(value) && setInputValue(text);
      onChangeText?.(text);
    }
  };

  /**
   * Handles press in animation for scale effect
   */
  const handlePressIn = () => {
    Animated.timing(scaleAnim, {
      duration: 100,
      toValue: 0.98,
      useNativeDriver: true,
    }).start();
  };

  /**
   * Handles press out animation to restore scale
   */
  const handlePressOut = () => {
    Animated.timing(scaleAnim, {
      duration: 100,
      toValue: 1,
      useNativeDriver: true,
    }).start();
  };

  useEffect(() => {
    setInputValue(value || '');
  }, [value]);

  return (
    <Box>
      {!!label && (
        <String
          size="sm"
          {...classNameHandler(twStyles.label)}
          {...styleHandler({ defaultStyle: styles.label })}
          {...labelStringProps}
        >
          {label}
        </String>
      )}
      <AnimatedBox
        ref={wrapperRef}
        {...styleHandler({
          defaultStyle: [inputContainerSv({ hasError: !!hasError, isDisabled: !!isDisabled })],
          style: [
            clickableAs === 'scale' ? { transform: [{ scale: scaleAnim }] } : undefined,
            hasError && errorStyle,
            style,
          ],
        })}
        {...classNameHandler(
          cn(
            inputContainerTv({ hasError: !!hasError, isDisabled: !!isDisabled }),
            hasError && errorClassName,
            className,
          ),
        )}
      >
        {!!leftIconAction?.iconName && !leftIconAction?.customIcon && (
          <Icon
            name={leftIconAction?.iconName as any}
            size={leftIconAction?.iconSize || xmarkIconSize}
            color={leftIconAction?.iconColor}
            onPress={leftIconAction?.iconPress}
            hasPressEffect={leftIconAction?.hasPressEffect}
            {...classNameHandler(cn(twStyles.leftIcon))}
            {...styleHandler({ defaultStyle: styles.leftIcon })}
          />
        )}
        {leftIconAction?.customIcon && (
          <Pressable
            onPress={leftIconAction?.iconPress}
            {...classNameHandler(cn(twStyles.leftIcon))}
            {...styleHandler({ defaultStyle: styles.leftIcon })}
          >
            {leftIconAction?.customIcon}
          </Pressable>
        )}

        <TextInput
          onPress={onPress}
          onPressIn={handlePressIn}
          onPressOut={handlePressOut}
          ref={ref}
          value={inputValue}
          {...classNameHandler(
            cn(inputFieldTv({ isStretchable: !!isStretchable, multiline: !!multiline, size }), inputClassName),
          )}
          {...styleHandler({ defaultStyle: [inputSv({ isStretchable, multiline, size })], style: inputStyle })}
          onChangeText={handleOnChange}
          secureTextEntry={isSecureEntry}
          multiline={multiline}
          maxLength={maxLength}
          editable={isDisabled ? false : !!editable}
          {...rest}
        />

        <Box
          {...classNameHandler(cn(twStyles.iconContainer))}
          {...styleHandler({ defaultStyle: styles.iconContainer })}
        >
          {inputValue.length > 0 && hasClearIcon && (
            <Icon onPress={handleClearInput} name="xmark-solid" size={xmarkIconSize} />
          )}

          {!!rightIconAction?.iconName && !rightIconAction?.customIcon && (
            <Icon
              name={rightIconAction?.iconName as any}
              size={rightIconAction?.iconSize || xmarkIconSize}
              color={rightIconAction?.iconColor}
              onPress={rightIconAction?.iconPress}
              hasPressEffect={rightIconAction?.hasPressEffect}
              {...classNameHandler(cn(twStyles.rightIcon))}
              {...styleHandler({ defaultStyle: styles.rightIcon })}
            />
          )}
          {rightIconAction?.customIcon && (
            <Pressable
              onPress={rightIconAction?.iconPress}
              {...classNameHandler(cn(twStyles.rightIcon))}
              {...styleHandler({ defaultStyle: styles.rightIcon })}
            >
              {rightIconAction?.customIcon}
            </Pressable>
          )}

          {hasSecureTextEntry && (
            <Icon
              onPress={() => setIsSecureEntry(!isSecureEntry)}
              name={isSecureEntry ? 'eye-slash-solid' : 'eye-solid'}
              size="sm"
            />
          )}
        </Box>
      </AnimatedBox>
      <Box
        {...classNameHandler(
          cn(twStyles.bottomInputContainer, {
            [twStyles.bottomInputContainerShowLength]: showLength && !errorMessage,
          }),
        )}
        {...styleHandler({
          defaultStyle: styles.bottomInputContainer,
          style: showLength && !errorMessage && { justifyContent: 'flex-end' },
        })}
      >
        {!!errorMessage && (
          <Box
            {...classNameHandler(cn(twStyles.errorContainer))}
            {...styleHandler({ defaultStyle: styles.errorContainer })}
          >
            {!!errorIconName && <Icon name={errorIconName as any} size="xs" color="#FF0000" />}
            <String
              size="xs"
              colorVariant="error"
              {...styleHandler({ defaultStyle: styles.errorText })}
              {...errorStringProps}
            >
              {errorMessage}
            </String>
          </Box>
        )}
        {showLength && (
          <String
            size="xs"
            {...classNameHandler(cn('mr-0.5'))}
            {...styleHandler({ defaultStyle: styles.lengthText })}
            {...lengthStringProps}
          >
            {maxLength ? `${getStringLength(inputValue ?? '')}/${maxLength}` : getStringLength(inputValue ?? '')}
          </String>
        )}
      </Box>
    </Box>
  );
});

Input.displayName = 'Input';

export default Input;
