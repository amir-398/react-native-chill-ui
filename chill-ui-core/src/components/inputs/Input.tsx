import { cn } from '@utils';
import { forwardRef, useEffect, useState } from 'react';
// eslint-disable-next-line
import Animated, { useAnimatedStyle, useSharedValue, withTiming } from 'react-native-reanimated';
import { Pressable, TextInput, View } from 'react-native';

import Icon from '../icon';
import { Box } from '../box';
import { String } from '../string';
import inputSizeVariants from './styleVariants';
import { InputProps } from '../../types/input.types';
import { getStringLength, isString } from '../../utils';

/**
 * Input component with advanced features including validation, icons, animations, and accessibility support.
 * Supports multiple input types, validation patterns, icons, error handling, and customizable styling.
 *
 * @example
 * ```tsx
 * // Basic input
 * <Input
 *   placeholder="Enter your name"
 *   onChangeText={(text) => console.log(text)}
 * />
 *
 * // Input with label and validation
 * <Input
 *   label="Email Address"
 *   placeholder="Enter your email"
 *   allow="all"
 *   hasError={!!emailError}
 *   errorMessage={emailError}
 *   onChangeText={setEmail}
 * />
 *
 * // Input with icons
 * <Input
 *   placeholder="Search..."
 *   leftIconAction={{
 *     iconName: 'search-solid',
 *     iconSize: 'sm',
 *     iconColor: '#666',
 *   }}
 *   rightIconAction={{
 *     iconName: 'filter-solid',
 *     iconSize: 'sm',
 *     iconPress: handleFilter,
 *   }}
 *   onChangeText={setSearchQuery}
 * />
 *
 * ```
 *

 * @param label - Label text for the input
 * @param className - Custom CSS classes for the input container
 * @param labelClassName - Custom CSS classes for the label
 * @param wrapperRef - Reference to the wrapper component
 * @param hasError - Whether the input has an error
 * @param errorMessage - Error message to display
 * @param errorClassName - Custom CSS classes for the error message
 * @param errorIconName - Icon name to display with error
 * @param hasClearIcon - Whether to show clear icon
 * @param inputClassName - Custom CSS classes for the input field
 * @param leftIconAction - Left icon configuration with iconName, iconColor, iconSize, customIcon, iconPress
 * @param rightIconAction - Right icon configuration with iconName, iconColor, iconSize, customIcon, iconPress
 * @param hasSecureTextEntry - Whether to show secure text entry
 * @param clickableAs - Type of clickable interaction ('pressable' | 'scale')
 * @param showLength - Whether to show character count
 * @param customRegex - Custom regex pattern for validation
 * @param allow - Allowed input types ('all' | 'numbers' | 'letters' | 'lettersWithoutSpecialCharacters')
 * @param isDisabled - Whether the input is disabled
 * @param isStretchable - Whether the input stretches to fill container
 * @param size - Size variant for the input ('xs' | 'sm' | 'md' | 'lg')
 * @param value - Current input value
 * @param onChangeText - Callback when text changes
 * @param placeholder - Placeholder text
 * @param multiline - Whether input supports multiple lines
 * @param maxLength - Maximum character length
 * @param editable - Whether input is editable
 * @param onPress - Callback when input is pressed
 * @param secureTextEntry - Whether to show secure text entry (inherited from TextInputProps)
 * @param ref - Forwarded ref to the underlying TextInput component
 * @returns Input component with label, validation, icons, and error handling
 */
const Input = forwardRef<TextInput, InputProps>((props, ref) => {
  const {
    allow = 'all',
    className,
    clickableAs = 'scale',
    customRegex,
    editable = true,
    errorClassName,
    errorIconName,
    errorMessage,
    hasClearIcon = true,
    hasError,
    hasSecureTextEntry,
    inputClassName,
    isDisabled,
    isStretchable,
    label,
    labelClassName,
    leftIconAction,
    maxLength,
    multiline,
    onChangeText,
    onPress,
    rightIconAction,
    showLength,
    size = 'md',
    value,
    wrapperRef,
    ...rest
  } = props;

  const [isSecureEntry, setIsSecureEntry] = useState<boolean>(hasSecureTextEntry || false);
  const [inputValue, setInputValue] = useState<string>(value || '');
  const scale = useSharedValue(1);

  /**
   * Validates input text based on allow prop or custom regex
   * @param text - Text to validate
   * @returns Whether the text is valid according to validation rules
   */
  const validateInput = (text: string): boolean => {
    if (customRegex) {
      return customRegex.test(text);
    }

    switch (allow) {
      case 'numbers':
        return /^\d*$/.test(text);
      case 'letters':
        return /^[a-zA-ZÀ-ÿ\s]*$/.test(text);
      case 'lettersWithoutSpecialCharacters':
        return /^[a-zA-ZÀ-ÿ0-9\s]*$/.test(text);
      case 'all':
      default:
        return true;
    }
  };

  /**
   * Clears the input value and calls onChangeText callback
   */
  const handleClearInput = () => {
    setInputValue('');
    onChangeText?.('');
  };

  /**
   * Handles text changes with validation and state updates
   * @param text - New text value to process
   */
  const handleOnChange = (text: string) => {
    if (validateInput(text)) {
      !isString(value) && setInputValue(text);
      onChangeText?.(text);
    }
  };

  const animatedStyle = useAnimatedStyle(() => ({
    transform: [{ scale: scale.value }],
  }));

  /**
   * Handles press in animation for scale effect
   */
  const handlePressIn = () => {
    scale.value = withTiming(0.98, { duration: 100 });
  };

  /**
   * Handles press out animation to restore scale
   */
  const handlePressOut = () => {
    scale.value = withTiming(1, { duration: 100 });
  };

  useEffect(() => {
    setInputValue(value || '');
  }, [value]);

  /**
   * Renders the TextInput component with applied styles and props
   * @returns TextInput component with proper styling and behavior
   */
  const renderInput = () => (
    <TextInput
      onPress={onPress}
      onPressIn={handlePressIn}
      onPressOut={handlePressOut}
      ref={ref}
      value={inputValue}
      className={cn(
        'flex-1',
        inputSizeVariants({ isStretchable, multiline, size }),
        inputClassName,
        hasError && errorClassName,
      )}
      onChangeText={handleOnChange}
      secureTextEntry={isSecureEntry}
      multiline={multiline}
      maxLength={maxLength}
      editable={isDisabled ? false : !!editable}
      {...rest}
    />
  );

  /**
   * Renders the input with optional pressable wrapper for animations
   * @returns Input component with or without animated pressable wrapper
   */
  const renderPressableInput = () => {
    if (clickableAs === 'scale') {
      return (
        <Animated.View
          style={animatedStyle}
          className={cn('flex-1', inputSizeVariants({ isStretchable, multiline, size }))}
        >
          {renderInput()}
        </Animated.View>
      );
    }
    return renderInput();
  };

  return (
    <Box>
      {label && (
        <String size="sm" className={cn('mb-1 ml-1', labelClassName)}>
          {label}
        </String>
      )}
      <View
        ref={wrapperRef}
        className={cn(
          'flex flex-row items-center rounded-lg border border-gray-300 bg-white px-3 text-black',
          { 'border-red-500': hasError },
          { 'opacity-50': isDisabled },
          className,
        )}
      >
        {!!leftIconAction?.iconName && !leftIconAction?.customIcon && (
          <Icon
            name={leftIconAction?.iconName}
            size={leftIconAction?.iconSize || 'sm'}
            color={leftIconAction?.iconColor}
            onPress={leftIconAction?.iconPress}
            className="mr-1.5"
          />
        )}
        {leftIconAction?.customIcon && (
          <Pressable onPress={leftIconAction?.iconPress} className="mr-1.5">
            {leftIconAction?.customIcon}
          </Pressable>
        )}

        {onPress ? renderPressableInput() : renderInput()}

        <Box className="flex-row items-center gap-1">
          {inputValue.length > 0 && hasClearIcon && <Icon onPress={handleClearInput} name="xmark-solid" size="sm" />}

          {!!rightIconAction?.iconName && !rightIconAction?.customIcon && (
            <Icon
              name={rightIconAction?.iconName}
              size={rightIconAction?.iconSize || 'sm'}
              color={rightIconAction?.iconColor}
              onPress={rightIconAction?.iconPress}
            />
          )}
          {rightIconAction?.customIcon && (
            <Pressable onPress={rightIconAction?.iconPress}>{rightIconAction?.customIcon}</Pressable>
          )}
        </Box>

        {hasSecureTextEntry && (
          <Icon
            onPress={() => setIsSecureEntry(!isSecureEntry)}
            name={isSecureEntry ? 'eye-slash-solid' : 'eye-solid'}
            size="sm"
          />
        )}
      </View>
      <Box className={cn('flex-row items-center justify-between gap-1', showLength && 'justify-end')}>
        {hasError && (
          <Box className="h-5 flex-row items-center gap-1 pl-1">
            {!!errorIconName && <Icon name={errorIconName} size="xs" color="#FF0000" />}
            <String size="xs" colorVariant="error">
              {errorMessage}
            </String>
          </Box>
        )}
        {showLength && (
          <String size="xs" className="mr-0.5">
            {maxLength ? `${getStringLength(inputValue ?? '')}/${maxLength}` : getStringLength(inputValue ?? '')}
          </String>
        )}
      </Box>
    </Box>
  );
});

Input.displayName = 'Input';

export default Input;
