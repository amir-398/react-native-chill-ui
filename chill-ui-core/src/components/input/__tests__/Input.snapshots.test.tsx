import { render } from '@testing-library/react-native';

import Input from '../components/Input';

// Mocks simples
jest.mock('../styles/Input.styles', () => ({
  inputContainerSv: jest.fn(() => ({})),
  inputSv: jest.fn(() => ({})),
  styles: {},
}));

jest.mock('../../../utils', () => ({
  classNameHandler: jest.fn(() => ({})),
  classNamePropsHandler: jest.fn(),
  cn: jest.fn((...args) => args.filter(Boolean).join(' ')),
  getStringLength: jest.fn(str => str?.length || 0),
  isString: jest.fn(value => typeof value === 'string'),
  styleHandler: jest.fn(() => ({})),
}));

jest.mock('../../../components/box', () => ({
  Box: ({ children }: any) => children,
}));

jest.mock('../../../components/icon', () => ({
  Icon: ({ name, onPress }: any) => {
    function MockIcon() {
      return <div data-testid={`icon-${name}`} onClick={onPress} />;
    }
    return <MockIcon />;
  },
}));

jest.mock('../../../components/string', () => ({
  String: ({ children }: any) => <div>{children}</div>,
}));

jest.mock('../../../components/animatedBox', () => ({
  AnimatedBox: ({ children }: any) => children,
}));

describe('Input Component Snapshots', () => {
  it('should match snapshot for basic input', () => {
    const tree = render(<Input placeholder="Basic input" />);
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot for input with error', () => {
    const tree = render(<Input placeholder="Input with error" hasError errorMessage="This is an error message" />);
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot for password input', () => {
    const tree = render(<Input placeholder="Enter password" hasSecureTextEntry value="password123" />);
    expect(tree).toMatchSnapshot();
  });
});
