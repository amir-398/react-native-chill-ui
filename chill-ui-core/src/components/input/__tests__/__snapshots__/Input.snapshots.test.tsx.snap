// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component Snapshots should match snapshot for basic input 1`] = `
<TextInput
  editable={true}
  onChangeText={[Function]}
  onPressIn={[Function]}
  onPressOut={[Function]}
  placeholder="Basic input"
  secureTextEntry={false}
  value=""
/>
`;

exports[`Input Component Snapshots should match snapshot for input with error 1`] = `
[
  <TextInput
    editable={true}
    onChangeText={[Function]}
    onPressIn={[Function]}
    onPressOut={[Function]}
    placeholder="Input with error"
    secureTextEntry={false}
    value=""
  />,
  <div>
    This is an error message
  </div>,
]
`;

exports[`Input Component Snapshots should match snapshot for password input 1`] = `
[
  <TextInput
    editable={true}
    onChangeText={[Function]}
    onPressIn={[Function]}
    onPressOut={[Function]}
    placeholder="Enter password"
    secureTextEntry={true}
    value="password123"
  />,
  <div
    data-testid="icon-xmark-solid"
    onClick={[Function]}
  />,
  <div
    data-testid="icon-eye-slash-solid"
    onClick={[Function]}
  />,
]
`;
